Feature/spherical-joints

marosmester opened the pull request:
WORK IN PROGRESS

## What is the current behavior?
Only revolute joint type supported.

## What is the new behavior?
The goal is to add support for spherical joints.

aleskucera reviewed (changes requested):
This seems to be a good start. I added attribute `num_constraints` to the joint interaction, because I think you will need it. I also think that like this the file `joint_interaction.py` is finished. But right now it comes the hard part. You have to study, how exactly are the joint constraints enforced and I don't know if the changes you have made in `engine_dims.py` are in the right direction. But continue and if you want some more feedback, feel free to ask.

marosmester commented:
Thanks for the previous feedback. I switched up my strategy. As you have mentioned in person, I now also see that there are at least 2 ways to tackle this feature:

**A)** write `joint_interaction_kernel` and `joint_constraint_kernel` as global kernels for all joint types, thus handling types in the kernels and having the joint_constraint block full of mixed joint_types

**B)** write separate versions of  `joint_interaction_kernel` and `joint_constraint_kernel` for the 2 joint types. This sometimes results in larger code overhead. But IMO it makes the code cleaner by creating dedicated h, J, C arrays for each joint type. 

Up until a commit labeled [DEAD END] I followed the approach A but now switched to approach B. I can revert back to A if needed. Feedback would be appreciated but can wait for an in-person meeting.
